iOS/TvShows/TvShows/Common/Network/BaseApiClient.swift

```diff
@@ -0,0 +1,164 @@
+import Foundation
+import RxCocoa
+import RxSwift
+
+class BaseApiClient: ApiClientProtocol {
+
+    private let baseUrl: String
+    private let urlSession: URLSession
+
+    init(baseUrl: String) {
+        self.baseUrl = baseUrl
+        self.urlSession = URLSession.shared
+    }
+
+    func get<ResultType: Decodable>(
+        path: String,
+        queryParameters: [String: String]?
+    ) -> Single<ResultType> {
+        executeAndReturn(path: url(with: path), method: .get, parameters: queryParameters)
+    }
+
+    func post<ParamsType: Encodable, ResultType: Decodable>(
+        path: String,
+        body: ParamsType
+    ) -> Single<ResultType> {
+        executeAndReturn(path: url(with: path), method: .post, parameters: body)
+    }
+
+    private func executeAndReturn<ParamsType: Encodable, ResultType: Decodable>(
+        path: String,
+        method: HTTPMethod,
+        parameters: ParamsType? = nil
+    ) -> Single<ResultType> {
+        .create { [weak self] single in
+            guard let self = self else { return Disposables.create() }
+
+            let request = self.buildRequest(path: path, method: method, parameters: parameters)
+            self.urlSession.dataTask(with: request) { [weak self] data, response, error in
+                guard
+                    let httpReponse = response as? HTTPURLResponse,
+                    let self = self,
+                    let statusCode = HttpStatusCode(rawValue: httpReponse.statusCode)
+                else {
+                    single(.failure(ApiError.badRequest))
+                    return
+                }
+                
+                if let error = self.mapToApiError(status: statusCode) {
+                    single(.failure(error))
+                    return
+                }
+                
+                guard let result: ResultType = self.parse(data: data) else {
+                    single(.failure(ApiError.noData))
+                    return
+                }
+
+                single(.success(result))
+            }
+            .resume()
+            
+            return Disposables.create()
```

Comment:
This can be improved regarding the task cancelation. 
Also you don't even need that guard you have at the start of the function. 
If `self` is `nil` data task won't be created and you'll return Disposable at the end regardless of that.

Reply:
If I do not guard self, I have to guard the result of the method self.buildRequest() since self will be optional in that case and urlSession.dataTask(with: request) cannot take the optional request. Therefore, I end up with the same inconvenience of having to guard buildRequest() method result and writing it like this:

```swift
guard let request = self?.buildRequest(path: path, method: method, parameters: parameters) else {
                single(.failure(ApiError.badRequest))
                return Disposables.create()
            }
```